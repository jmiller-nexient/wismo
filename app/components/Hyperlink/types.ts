import React from 'react';

export type BrowserTarget = '_blank' | '_parent' | '_self' | '_top';

export interface IBaseHyperlinkProps {
  children?: React.ReactNode;
  href: string;
  target?: BrowserTarget;
}

export interface IHyperlinkProps extends IBaseHyperlinkProps {
  alt?: string | null;
}
