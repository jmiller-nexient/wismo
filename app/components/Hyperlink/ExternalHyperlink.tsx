/**
 * A link to a certain page, an anchor tag
 */
import React from 'react';
import { IHyperlinkProps } from './types';

const ExternalHyperlink: React.FC<IHyperlinkProps> = (props: IHyperlinkProps): React.ReactElement => {
  const { alt, children, href, target } = props;

  return (
    <a href={`${href}`} target={target || '_blank'}>
      <img
        alt={alt || 'external-link'}
        className="dte-wismo-hyperlink-icon"
        src={require('images/external-link.svg')}
      />
      {React.Children.toArray(children)}
    </a>
  );
};

export default ExternalHyperlink;
