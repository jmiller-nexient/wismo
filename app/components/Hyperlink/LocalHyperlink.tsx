/**
 * A link to a certain page, an anchor tag
 */
import React from 'react';
import { Link } from 'react-router-dom';
import { IBaseHyperlinkProps } from './types';

// tslint:disable-next-line: no-empty-interface
interface ILocalHyperlinkProps extends IBaseHyperlinkProps {}

const LocalHyperlink: React.FC<ILocalHyperlinkProps> = (props: ILocalHyperlinkProps): React.ReactElement => {
  const { children, href, target } = props;

  return (
    <Link to={`${href}`} target={target || '_self'}>
      {React.Children.toArray(children)}
    </Link>
  );
};

export default LocalHyperlink;
