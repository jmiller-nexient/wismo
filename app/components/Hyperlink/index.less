@import '../../styles/themes/index.less';

@hyperlink-prefix-cls: ~'@{dte-prefix}-hyperlink';

.@{hyperlink-prefix-cls}-icon {
  line-height: 16px;
  margin-right: 10px;
  width: 16px;
}
