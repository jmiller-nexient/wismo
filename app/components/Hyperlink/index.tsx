/**
 * A link to a certain page, an anchor tag
 */
import React from 'react';

import './index.less';

type Target = '_blank' | '_parent' | '_self' | '_top';

interface HyperlinkProps {
  alt?: string | null;
  children?: React.ReactNode;
  href: string;
  target?: Target;
}

const Hyperlink: React.FC<HyperlinkProps> = (props: HyperlinkProps): React.ReactElement => {
  const { alt, children, href, target } = props;

  const ExternalIcon =
    href && href.includes(window.location.hostname) ? null : (
      <img
        alt={alt || 'external-link'}
        className="dte-wismo-hyperlink-icon"
        src={require('images/external-link.svg')}
      />
    );

  return (
    <a href={`${href}`} target={target || (ExternalIcon ? '_blank' : '_self')}>
      {ExternalIcon}
      {React.Children.toArray(children)}
    </a>
  );
};

export default Hyperlink;
