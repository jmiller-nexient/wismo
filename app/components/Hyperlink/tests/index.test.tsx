import React from 'react';
import renderer from 'react-test-renderer';

import Hyperlink from '..';

const externalUrl = 'https://www.google.com';
const internalUrl = window.location.hostname;

describe('<Hyperlink />', () => {
  it('should render external link when href is for a different host', () => {
    const tree = renderer.create(
      <Hyperlink alt="external-icon" href={externalUrl}>External Link Text</Hyperlink>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render internal link when href is for the current host', () => {
    const tree = renderer.create(
      <Hyperlink alt="internal-link" href={internalUrl}>Internal Link Text</Hyperlink>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  describe('target', () => {
    it('should be set to _blank when a target is not provided and href is for a different host', () => {
      const tree = renderer.create(
        <Hyperlink alt="external-link" href={externalUrl}>External Link Text</Hyperlink>,
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should be set to _self when a target is not provided and href is for the current host', () => {
      const tree = renderer.create(
        <Hyperlink alt="internal-link" href={internalUrl}>Internal Link Text</Hyperlink>,
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should be set to the given target when a target is provided and href is for a different host', () => {
      const tree = renderer.create(
        <Hyperlink alt="external-link" href={externalUrl} target="_top">External Link Text</Hyperlink>,
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should be set to the given target when a target is provided and href is for the current host', () => {
      const tree = renderer.create(
        <Hyperlink alt="internal-link" href={internalUrl} target="_top">Internal Link Text</Hyperlink>,
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
