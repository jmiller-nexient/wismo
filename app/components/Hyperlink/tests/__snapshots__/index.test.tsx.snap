// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Hyperlink /> should render external link when href is for a different host 1`] = `
<a
  href="https://www.google.com"
  target="_blank"
>
  <img
    alt="external-icon"
    className="dte-wismo-hyperlink-icon"
    src="IMAGE_MOCK"
  />
  External Link Text
</a>
`;

exports[`<Hyperlink /> should render internal link when href is for the current host 1`] = `
<a
  href="localhost"
  target="_self"
>
  Internal Link Text
</a>
`;

exports[`<Hyperlink /> target should be set to _blank when a target is not provided and href is for a different host 1`] = `
<a
  href="https://www.google.com"
  target="_blank"
>
  <img
    alt="external-link"
    className="dte-wismo-hyperlink-icon"
    src="IMAGE_MOCK"
  />
  External Link Text
</a>
`;

exports[`<Hyperlink /> target should be set to _self when a target is not provided and href is for the current host 1`] = `
<a
  href="localhost"
  target="_self"
>
  Internal Link Text
</a>
`;

exports[`<Hyperlink /> target should be set to the given target when a target is provided and href is for a different host 1`] = `
<a
  href="https://www.google.com"
  target="_top"
>
  <img
    alt="external-link"
    className="dte-wismo-hyperlink-icon"
    src="IMAGE_MOCK"
  />
  External Link Text
</a>
`;

exports[`<Hyperlink /> target should be set to the given target when a target is provided and href is for the current host 1`] = `
<a
  href="localhost"
  target="_top"
>
  Internal Link Text
</a>
`;
